### PR TITLE
Implement account manage manage identities

### PR DESCRIPTION
**Pull request by cychiuae**

### CODE REVIEW

**Comment by cychiuae** [comment]
@louischan-oursky Please have a look first if you have time.

File: pkg/lib/accountmanagement/redis_store.go [hunk 1]
```diff
@@ -22,10 +22,20 @@ type RedisStore struct {
 }
 
 type GenerateTokenOptions struct {
+	// OAuth
 	UserID      string
 	Alias       string
 	MaybeState  string
 	RedirectURI string
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/redis_store.go:29]: Technically speaking, UserID is shared, and only Alias, MaybeState, RedirectURI are OAuth specific.

File: pkg/lib/accountmanagement/redis_store.go [hunk 2]
```diff
@@ -36,15 +46,26 @@ func (s *RedisStore) GenerateToken(options GenerateTokenOptions) (string, error)
 	ttl := duration.UserInteraction
 	expireAt := now.Add(ttl)
 
+	IdentityToken := &IdentityToken{
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/redis_store.go:49]: lowercase.

File: pkg/lib/accountmanagement/token.go [hunk 3]
```diff
@@ -9,22 +9,33 @@ import (
...
+	// Adding OAuth
+	Alias       string `json:"alias,omitempty"`
+	RedirectURI string `json:"redirect_uri,omitempty"`
+	State       string `json:"state,omitempty"`
+
+	// Adding Identity
+	Identity *IdentityToken `json:"identity_token,omitempty"`
+}
+
+type IdentityToken struct {
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/token.go:27]: I prefer we name this TokenIdentity

File: pkg/lib/accountmanagement/token.go [hunk 4]
```diff
@@ -9,22 +9,33 @@ import (
...
+	CreatedAt *time.Time `json:"created_at,omitempty"`
+	ExpireAt  *time.Time `json:"expire_at,omitempty"`
+
+	// Adding OAuth
+	Alias       string `json:"alias,omitempty"`
+	RedirectURI string `json:"redirect_uri,omitempty"`
+	State       string `json:"state,omitempty"`
+
+	// Adding Identity
+	Identity *IdentityToken `json:"identity_token,omitempty"`
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/token.go:24]: `token_identity,omitempty`

File: pkg/lib/accountmanagement/token.go [hunk 5]
```diff
@@ -9,22 +9,33 @@ import (
...
+	Email       string `json:"email,omitempty"`
 }
 
 func (t *Token) CheckUser(userID string) error {
 	if subtle.ConstantTimeCompare([]byte(t.UserID), []byte(userID)) == 1 {
 		return nil
 	}
 
-	return ErrOAuthTokenNotBoundToUser
+	return ErrAccountManagementTokenNotBoundToUser
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/token.go:38]: This line changes the API accidentally. We should keep the existing OAuth API to return the same error.

We can keep the compatibility by adding CheckUser_OAuth and CheckUser. Change the OAuth code to use CheckUser_OAuth.

File: pkg/lib/accountmanagement/service_identity.go [hunk 6]
```diff
@@ -0,0 +1,1333 @@
...
+
+type AddIdentityBiometricInput struct {
+	JWTToken string
+}
+
+type AddIdentityBiometricOutput struct {
+	IdentityInfo *identity.Info
+}
+
+func (s *Service) AddIdentityBiometric(resolvedSession session.ResolvedSession, input *AddIdentityBiometricInput) (*AddIdentityBiometricOutput, error) {
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service_identity.go:971]: I do not think we support adding biometric identity in the settings page. Remove code related to adding biometric identity.

File: pkg/lib/accountmanagement/service.go [hunk 7]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
 
 	return &FinishAddingOutput{}, nil
 }
+
+type ResendOTPCodeInput struct {
+	UserID       string
+	Channel      model.AuthenticatorOOBChannel
+	LoginID      string
+	isSwitchPage bool
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:285]: This is a private field, and I think input should not have private fields. Anything wrong here?

File: pkg/lib/accountmanagement/service.go [hunk 8]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
 
 	return &FinishAddingOutput{}, nil
 }
+
+type ResendOTPCodeInput struct {
+	UserID       string
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:282]: It should not take UserID. It should take token. ResendOTPCode should be in the middleware of a flow.

File: pkg/lib/accountmanagement/service.go [hunk 9]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+	Channel      model.AuthenticatorOOBChannel
+	LoginID      string
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:284]: It should not take LoginID. The target to resend to should be stored in the token already.

File: pkg/lib/accountmanagement/service.go [hunk 10]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+	if err != nil {
+		return nil, err
+	}
+	return &ChangePasswordOutput{RedirectURI: redirectURI}, nil
+
+}
+
+type CreateAdditionalPasswordInput struct {
+	NewAuthenticatorID string
+	UserID             string
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:379]: No userID. Instead

```
CreateAdditionalPassword(resolvedSession, input)

type CreateAdditionalPasswordInput struct {
  PlainPassword string
}
```

File: pkg/lib/accountmanagement/service.go [hunk 11]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ChangePasswordOutput{RedirectURI: redirectURI}, nil
+
+}
+
+type CreateAdditionalPasswordInput struct {
+	NewAuthenticatorID string
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:378]: No NewAuthenticatorID. It is not an information the caller should care about.

File: pkg/lib/accountmanagement/service.go [hunk 12]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+	Password           string
+}
+
+func NewCreateAdditionalPasswordInput(userID string, password string) CreateAdditionalPasswordInput {
+	return CreateAdditionalPasswordInput{
+		NewAuthenticatorID: uuid.New(),
+		UserID:             userID,
+		Password:           password,
+	}
+}
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:389]: Remove this function. CreateAdditionalPasswordInput is simple struct that does not need a function to initialize.

File: pkg/lib/accountmanagement/service.go [hunk 13]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+
+func (s *Service) CreateAdditionalPassword(input CreateAdditionalPasswordInput) error {
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:391]: It takes `resolvedSession, input`

File: pkg/lib/accountmanagement/service.go [hunk 14]
```diff
@@ -184,3 +277,220 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
+		},
+	}
+	info, err := s.Authenticators.NewWithAuthenticatorID(input.NewAuthenticatorID, spec)
+	if err != nil {
+		return err
+	}
+	return s.CreateAuthenticator(info)
+}
+
+func (s *Service) CreateAuthenticator(authenticatorInfo *authenticator.Info) error {
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:408]: This function should be private.

File: pkg/lib/accountmanagement/service_identity.go [hunk 15]
```diff
@@ -0,0 +1,1333 @@
...
+	LoginIDKey string
+	Token      string
+	Code       string
+}
+
+type ResumeAddIdentityEmailOutput struct {
+	IdentityInfo *identity.Info
+}
+
+func (s *Service) ResumeAddIdentityEmail(resolvedSession session.ResolvedSession, input *ResumeAddIdentityEmailInput) (output *ResumeAddIdentityEmailOutput, err error) {
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service_identity.go:241]: I would like to come up with a convention

- If it is a one-step operation, then the signature should be `(resolvedSession, input)`
- If it is the start of a multi-step operation, then it should be `StartXXX(resolvedSession, input)`
- If it is the continuation of a multi-step operation, then it should be `ResumeXXX(resolvedSession, token, input)` or `FinishXXX(resolvedSession, token, input)`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/accountmanagement/service_identity.go (hunk 6 above)
Comment (cychiuae) [reply pkg/lib/accountmanagement/service_identity.go:971]: oh right

File: pkg/lib/accountmanagement/token.go (hunk 5 above)
Comment (cychiuae) [reply pkg/lib/accountmanagement/token.go:38]: We've kept the api by checking err if `ErrAccountManagementTokenNotBoundToUser` and then rethrow `ErrOAuthTokenNotBoundToUser`.

Anyway, let me change to implement `CheckUser_OAuth` instead

File: pkg/lib/accountmanagement/service.go [hunk 16]
```diff
@@ -270,20 +271,41 @@ func (s *Service) FinishAdding(input *FinishAddingInput) (*FinishAddingOutput, e
...
-	isResend := !input.isSwitchPage
+func (s *Service) ResendOTPCode(resolvedSession session.ResolvedSession, input *ResendOTPCodeInput) (err error) {
+	userID := resolvedSession.GetAuthenticationInfo().UserID
+
+	token, err := s.Store.GetToken(input.Token)
+	if err != nil {
+		return err
+	}
+
+	err = token.CheckUser_OAuth(userID)
```
Comment (louischan-oursky) [inline pkg/lib/accountmanagement/service.go:269]: Use CheckUser instead of CheckUser_OAuth

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by louischan-oursky** [comment]
You may need rebase.

**Comment by louischan-oursky** [comment]
@cychiuae or @IniZio

**Comment by IniZio** [comment]
@cychiuae Can give me push access to this branch? 🙏

**Comment by IniZio** [comment]
@louischan-oursky Rebased

**Comment by louischan-oursky** [comment]
@IniZio You need to run `make generate` again.

**Comment by IniZio** [comment]
@louischan-oursky Done 🙏

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**